jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndex.java

```diff
@@ -152,6 +158,70 @@ private List<Int2ObjectHashMap<int[]>> loadInMemoryLayers(RandomAccessReader in)
         return imn;
     }
 
+    private Int2ObjectHashMap<FusedFeature.InlineSource> getInMemoryFeatures(RandomAccessReader in) throws IOException {
+        return inMemoryFeatures.updateAndGet(current -> {
+            if (current != null) {
+                return current;
+            }
+            // Only load the in-memory features if the graph is fused
+            for (var feature : features.values()) {
+                if (feature.isFused()) {
+                    try {
+                        return loadInMemoryFeatures(in);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }
+            }
+            return null;
+        });
+    }
+
+    private Int2ObjectHashMap<FusedFeature.InlineSource> loadInMemoryFeatures(RandomAccessReader in) throws IOException {
+        Int2ObjectHashMap<FusedFeature.InlineSource> hierarchyFeatures = new Int2ObjectHashMap<>();
+
+        long L0size = idUpperBound * (inlineBlockSize + Integer.BYTES * (1L + 1L + layerInfo.get(0).degree));
```

Comment:
I suggest we use `layerInfo.get(0).size` since `idUpperBound` is not necessarily correct here and we already have files written with invalid values there, but the size is correct in all cases.